src/a2a/server/apps/rest/rest_app.py

```diff
@@ -0,0 +1,237 @@
+import functools
+import json
+import logging
+import traceback
+
+from collections.abc import AsyncGenerator, AsyncIterator, Awaitable, Callable
+from typing import Any
+
+from pydantic import ValidationError
+from sse_starlette.sse import EventSourceResponse
+from starlette.requests import Request
+from starlette.responses import JSONResponse
+
+from a2a.server.apps.jsonrpc import (
+    CallContextBuilder,
+    DefaultCallContextBuilder,
+)
+from a2a.server.context import ServerCallContext
+from a2a.server.request_handlers.request_handler import RequestHandler
+from a2a.server.request_handlers.rest_handler import (
+    RESTHandler,
+)
+from a2a.types import (
+    A2AError,
+    AgentCard,
+    InternalError,
+    InvalidRequestError,
+    JSONParseError,
+    UnsupportedOperationError,
+)
+from a2a.utils.errors import MethodNotImplementedError
+
+
+logger = logging.getLogger(__name__)
+
+
+class RESTApplication:
+    """Base class for A2A REST applications.
+
+    Defines REST requests processors and the routes to attach them too, as well as
+    manages response generation including Server-Sent Events (SSE).
+    """
+
+    def __init__(
+        self,
+        agent_card: AgentCard,
+        http_handler: RequestHandler,
+        context_builder: CallContextBuilder | None = None,
+    ):
+        """Initializes the RESTApplication.
+
+        Args:
+            agent_card: The AgentCard describing the agent's capabilities.
+            http_handler: The handler instance responsible for processing A2A
+              requests via http.
+            context_builder: The CallContextBuilder used to construct the
+              ServerCallContext passed to the http_handler. If None, no
+              ServerCallContext is passed.
+        """
+        self.agent_card = agent_card
+        self.handler = RESTHandler(
+            agent_card=agent_card, request_handler=http_handler
+        )
+        self._context_builder = context_builder or DefaultCallContextBuilder()
+
+    def _generate_error_response(self, error: A2AError) -> JSONResponse:
+        """Creates a JSONResponse for an error.
+
+        Logs the error based on its type.
+
+        Args:
+            error: The Error object.
+
+        Returns:
+            A `JSONResponse` object formatted as a JSON error response.
+        """
+        log_level = (
+            logging.ERROR
+            if isinstance(error, InternalError)
+            else logging.WARNING
+        )
+        logger.log(
+            log_level,
+            'Request Error: '
+            f"Code={error.root.code}, Message='{error.root.message}'"
+            f'{", Data=" + str(error.root.data) if error.root.data else ""}',
+        )
+        return JSONResponse(
+            f'{{"message": "{error.root.message}"}}',
```

Comment:
No need to manually construct a JSON string here. I think `JSONResponse(error.model_dump())` would work.